src/riak_kv_pb_timeseries.erl

```diff
@@ -255,6 +261,108 @@ wait_until_active(Table, State, Seconds) ->
 %% functions called from check_table_and_call, one per ts* request
 %% ---------------------------------------------------
 
+
+%%
+%% INSERT statements, called from check_table_and_call.
+%%
+-spec make_insert_response(module(), #riak_sql_insert_v1{}) ->
+                           #tsqueryresp{} | #rpberrorresp{}.
+make_insert_response(Mod, #riak_sql_insert_v1{'INSERT' = Table, fields = Fields, values = Values}) ->
+    case lookup_field_positions(Mod, Fields) of
+	{error, FieldReason} ->
+	    make_rpberrresp(?E_BAD_QUERY, FieldReason);
+	{ok, Positions} ->
+        Empty = make_empty_row(Mod),
+        case xlate_insert_to_putdata(Values, Positions, Empty) of
+            {error, ValueReason} ->
+                make_rpberrresp(?E_BAD_QUERY, ValueReason);
+            {ok, Data} ->
+                Response = sub_putreq_common(Mod, Table, Data, #state{}),
+                xlate_put_to_qry_resp(Response)
+	    end
+    end.
+
+xlate_put_to_qry_resp({reply, #tsputresp{}, _State}) ->
+    #tsqueryresp{};
+xlate_put_to_qry_resp({reply, Err, _State}) ->
+    Err.
+
+
+%%
+%% Return an all-null empty row ready to be populated by the values
+%%
+-spec make_empty_row(module()) -> tuple(undefined).
+make_empty_row(Mod) ->
+    Positions = Mod:get_field_positions(),
+    list_to_tuple(lists:duplicate(length(Positions), undefined)).
+
+%%
+%% Lookup the index of the field names selected to insert.
+%%
+%% This *requires* that once schema changes take place the DDL fields are left in order.
+%%
+-spec lookup_field_positions(module(), [identifier()]) ->
+                           {ok, [pos_integer()]} | {error, string()}.
+lookup_field_positions(Mod, FieldIdentifiers) ->
+    case lists:foldl(
+	   fun({identifier, FieldName}, {Good, Bad}) ->
+		   case Mod:is_field_valid(FieldName) of
+               false ->
+                   {Good, [flat_format("undefined field ~s", [FieldName]) | Bad]};
+               true ->
+                   {[Mod:get_field_position(FieldName) | Good], Bad}
+		   end
+	   end, {[], []}, FieldIdentifiers)
+    of
+	{Positions, []} ->
+	    {ok, lists:reverse(Positions)};
+	{_, Errors} ->
+	    %% Only returns the first error, could investigate returning multiple.
```

Comment:
I suggest collecting just the bad fields in the foldl, and then formatting those for error message in a single `flat_format("undefined fields: ~s", [string:join(Fields, ", ")])`.